torch/csrc/api/src/optim/adam.cpp

```diff
@@ -12,57 +12,173 @@
 
 namespace torch {
 namespace optim {
-AdamOptions::AdamOptions(double learning_rate)
-    : learning_rate_(learning_rate) {}
+AdamOptions::AdamOptions(double lr) : lr_(lr) {}
+
+bool operator==(const AdamOptions& lhs, const AdamOptions& rhs) {
+  return (lhs.lr() == rhs.lr()) &&
+         (std::get<0>(lhs.betas()) == std::get<0>(rhs.betas())) &&
+         (std::get<1>(lhs.betas()) == std::get<1>(rhs.betas())) &&
+         (lhs.eps() == rhs.eps()) &&
+         (lhs.weight_decay() == rhs.weight_decay() &&
+         (lhs.amsgrad() == rhs.amsgrad()));
+}
+
+void AdamOptions::serialize(torch::serialize::OutputArchive& archive) const {
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(lr);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(betas);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(eps);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(weight_decay);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(amsgrad);
+}
+
+void AdamOptions::serialize(torch::serialize::InputArchive& archive) {
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, lr);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(betas_t, betas);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, eps);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, weight_decay);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(bool, amsgrad);
+}
+
+bool operator==(const AdamParamState& lhs, const AdamParamState& rhs) {
+  bool is_max_exp_avg_sq_defined_and_equal = !(lhs.max_exp_avg_sq().defined() ^ rhs.max_exp_avg_sq().defined());
+  if(!is_max_exp_avg_sq_defined_and_equal) {
+    is_max_exp_avg_sq_defined_and_equal = torch::equal(lhs.max_exp_avg_sq(), rhs.max_exp_avg_sq());
+  }
```

Comment:
When `lhs.max_exp_avg_sq()` is undefined and `rhs.max_exp_avg_sq()` is defined, `!(lhs.max_exp_avg_sq().defined() ^ rhs.max_exp_avg_sq().defined())` gives `false`, and `! is_max_exp_avg_sq_defined_and_equal` is `true`, which means we would go into the if branch and run `torch::equal` on undefined tensors (!). It might be easier to read and check for correctness if we write it as:
```cpp
  return (lhs.step() == rhs.step()) &&
          torch::equal(lhs.exp_avg(), rhs.exp_avg()) &&
          torch::equal(lhs.exp_avg_sq(), rhs.exp_avg_sq()) &&
          ((!lhs.max_exp_avg_sq().defined() && !rhs.max_exp_avg_sq().defined()) ||
           (lhs.max_exp_avg_sq().defined() && rhs.max_exp_avg_sq().defined() && torch::equal(lhs.max_exp_avg_sq(), rhs.max_exp_avg_sq())));
```